### DISCUSSION TITLE
Unable to visit url with a query parameter that shares the same slug name

### DISCUSSION THREAD

**Discussion by alex-r89** [vercel/next.js#48303] [Help]
### Summary

I have an application with a search route, located at `/search`

When a user searches for a term, I want them to be taken to the following url: `/search/?q=SEARCH_TERM`.

My understanding is that this would be structured with two files inside  the `/pages/search` folder of my application, these being `/pages/search/index.tsx` and `/pages/search/[q].tsx`

The issue I am facing is that I am unsure how to redirect them to this url, and run the code in the `[q].tsx` file. When a user searches, the code bellow takes them to the following url: `/search?q=SEARCH_TERM` - note is does not contain the second slash. 

Coupled with that, the user never leaves the `index.tsx` file - the code inside the `[q].tsx` file is never run as this is never hit by the user.

How do I correctly do this so that a user is taken to `/search/?q=SEARCH_TERM` with nextjs router?

Thanks

### Additional information

```js
router.push({
  pathname: 'search/',
  query: { q: searchTerm }
})
```

**Comment by ciruz** [comment]
Hmm no you can have it in the same file, I made a quick and dirty example:

```javascript
// pages/search.js

import React, { useEffect, useRef, useState } from "react";

import { useRouter } from "next/router";

export default function Search() {
  const router = useRouter();

  const [searchTerm, setSearchTerm] = useState("");

  const inputRef = useRef(null);

  // set the search term when the router query changes (like initial loading, juist try to remove it and see the difference when you reload the page with a ?q set)
  useEffect(() => {
    if (router.query.q) setSearchTerm(router.query.q);
  }, [router]);

  const handleFormSubmit = (event) => {
    event.preventDefault(); // prevent default form submission behavior

    const searchTerm = inputRef.current.value.trim(); // get the search term from the input field
    if (searchTerm !== "") {
      router.push({
        pathname: "/search",
        query: { q: searchTerm },
      });
      setSearchTerm(searchTerm);
    }
  };

  return (
    <div>
      <form onSubmit={handleFormSubmit}>
        <input type="text" name="q" ref={inputRef} defaultValue={searchTerm} />{" "}
        <button type="submit">submit</button>
      </form>
      <br />
      {searchTerm}
    </div>
  );
}
```

![image](https://user-images.githubusercontent.com/987947/231543645-0f5660a0-a1b5-4a8b-8ae9-213b7977b4e8.png)


The ``[q].js`` file, is a dynamic segment approach, this would mean the url needs to look like this ``/search/foobar``.

Then you would have two files `/pages/search.js` and `/pages/search/[q].js`.

Your `[q].js` file would look something like this:

```javascript
//[q].js
import { useRouter } from 'next/router'

const Page = () => {
  const router = useRouter()
  const { q } = router.query

  return <p>Term: {q}</p>
}

export default Page 
```

**Reply by alex-r89** [reply]
Hi,

Thanks for the reply! I think I may have explained myself poorly though:

If you take your example and go to `http://localhost:3000/search/?q=term`, you will not hit the `[q].js` file - this is what I would like to solve.

I understand this can all be done inside one file, but this is not what I want.

I may have this set up wrong/configured wrong/coded wrong but I specifically want the user to be directed to `/search/?q=SEARCH_TERM` and I specifically want the code inside `[q].tsx` (or whatever it should be named for this to work) to be run when a user visits the url `http://localhost:3000/search/?q=term`

EDIT:
Just to note, if i navigate directly to `http://localhost:3000/search/?q=term` then the slash after search is removed, ie the url is update quickly to be `http://localhost:3000/search?q=term`

I think this may be the issue. How i stop this from happening I am unsure as of yet

**Reply by ciruz** [reply]
Hey, before we dig further, there is a way to globally enable the [trailingSlash](https://nextjs.org/docs/api-reference/next.config.js/trailing-slash) via the `next.config.js`. **All** your urls are set to this `/foobar/` instead of `/foobar`.

Would it then be as you want it?

If not, we can work on a solution with the file.

**Reply by alex-r89** [reply]
Hi,

Thanks for that. I have added this via the next config. This does seem to change the behaviour, but it does not solve the issue:

With the trailing slash option enabled in the next.config.js:

Navigating to `http://localhost:3000/search/?q=term` maintains the final slash (the one after the word `search`, however the `[q].tsx` is still never hit.

I have also tried naming this `[...q].tsx` but still when i navigate to the url I just mentioned, the file is never hit. Instead `/pages/search/index.tsx` is hit.

I dont seem to be able to hit `/pages/search/[q].tsx` or `/pages/search/[...q].tsx` when visiting `http://localhost:3000/search/?q=term`

**Reply by ciruz** [reply]
It's a GET-parameter if you have it like this.

To hit the q file you need to have your url like so:

`/search/term` or `/search/foobar`, then you hit your file and your `q` will be `term` or `foobar`.

**Reply by alex-r89** [reply]
I must be explaining myself poorly or some wires are getting crossed. The issue i am facing is regarding doing this with a query string, not just directing to /search/foobar. **To clarify: i want to navigate to `/search/?q=foobar`** - the issue is regarding the `?q=`, ie the query string. 

I seem to have made some progress and worked out why this issue is happening. To step through it:

I have updated my folder structure to be as follows:

```js
/pages/search/[[...q]].tsx
/pages/search.tsx
```
However, This results in an error: 

`Error: You cannot define a route with the same specificity as a optional catch-all route`

This is understandable as the catch all (`[[...q]].tsx`) is conflicting with `search.tsx`

If I therefore rename `search.tsx` to be something else, lets say `hello.tsx` I am now finding that navigating to `http://localhost:3000/search/?q=term` works - the `[[...q]].tsx` is hit and the content from that file is served.

So there seems to be some issue here, either with my implementation or possibly with nextjs, regarding this setup I am trying to achieve.
It seems like you are not able to hit a url with a query parameter under a slug of the same name

**Reply by ciruz** [reply]
No, no you are not explaining yourself poorly, I totally understand what you want, I just tried to guide you a little, because I didn't understand why you cannot just enable `trailingSlashes` and use one file for it.

I'm almost sure what you want is not possible, even with next 13 and the new `/app` directory structure.
